src/skirmishbunker/FlatRoof.py

```diff
@@ -1,59 +1,338 @@
+        length -= 2 * self.roof_chamfer
+        return length
+
+    def calc_hatch_space_width(self):
+        width = self.width
+        width -= 2 * self.inset
+        width -= 2 * self.roof_chamfer
+        return width
+
+    def calc_hatch_length_offset(self):
+        return self.panel_length - self.hatch_length + self.panel_padding * 2
+
+    def calc_hatch_z_translate(self):
+        return (self.height / 2 + self.hatch_height / 2)
+
+    def calc_hole_x_translate(self):
+        translate = ((self.length - (self.inset * 2)) / 2)
+        translate -= (self.hole_diameter / 2)
+        translate -= self.hole_inset
+        return translate
+
+    def calc_hole_y_translate(self):
+        translate = ((self.width - (self.inset * 2)) / 2)
+        translate -= (self.hole_diameter / 2)
+        translate -= self.hole_inset
+        return translate
+
+    def calc_hole_z_translate(self):
+        return -1 * (self.height / 2 - self.hole_depth / 2)
+
+    # Calculate how far the default
+    # slot translation will be (This is ~85% of
+    # the tile size)
+    def calc_slot_translation(self):
+        return self.tile_size * 0.143
+    
+    # Calculate the slot radius/thickness.
+    # This is ~90.5% of the tile size
+    def calc_slot_radius(self):
+        return self.tile_size * 0.095
+    
+    # Calculate the length of mid size slots.
+    # This is ~66.666% of the tile size
+    def calc_slot_length_md(self):
+        return self.tile_size - (self.tile_size * 0.333)
+    
+    # Calculate the length of short size slots.
+    # This is ~33.333% of the tile size
+    def calc_slot_length_sm(self):
+        return self.tile_size - (self.tile_size * 0.666)
+
+    def make_roof_body(self):
+        roof_body = cq.Workplane("XY").box(
+            self.calc_final_length(),
+            self.calc_final_width(),
+            self.height
+        )
+
+        # if no chamfer/fillet, just return the new body
+        if self.roof_chamfer == 0:
+            self.roof_body = roof_body
+            return
+
+        # better to kill if this condition is met otherwise it fails silently.
+        if self.roof_chamfer >= self.height:
             raise Exception(f"roof_chamfer {self.roof_chamfer} >= roof height {self.height}")
 
+        if self.roof_operation.lower() == "chamfer":
+            roof_body = (
+                roof_body
+                .faces(self.roof_chamfer_faces_selector)
+                .edges(self.roof_chamfer_edges_selector)
+                .chamfer(self.roof_chamfer)
+            )
+        elif self.roof_operation.lower() == "fillet":
+            roof_body = (
+                roof_body
+                .faces(self.roof_chamfer_faces_selector)
+                .edges(self.roof_chamfer_edges_selector)
+                .fillet(self.roof_chamfer)
+            )
+        else:
+            raise Exception("Unrecognied roof operation")
+
         self.roof_body = roof_body
 
+    def make_tiles(self):
```

Comment:
As noted in previous dicussion this code will get refactored, to use the cqterrain tile method.
https://github.com/medicationforall/cqterrain/blob/main/src/cqterrain/tile/slotDiagonal.py
I'll do that on the next pass.

Reply:
I was going to open an issue about this separately. I think quite a lot of code could probably be decoupled and then rely on a factory or something 🤔